### ISSUE TITLE
[no-unnecessary-type-assertion] False negative if type is temporarily narrowed but narrowing is never relied on

### ISSUE THREAD

**Issue by peterflynn** [typescript-eslint/typescript-eslint#2953]
This is probably more of an enhancement request...

- [x] I have tried restarting my IDE and the issue persists.
- [ ] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```TS
let x: string | undefined = Math.random() > 0.5 ? "yes" : undefined;
let y: string | undefined = undefined;

if (Math.random() > 0.5) {
    y = x!;
}

if (y !== undefined) {
    console.log("Not undefined");
}
```

**Expected Result**

Expected:
"This assertion is unnecessary since the receiver accepts the original type of the expression."

By all appearances, the `!` postfix is unnecessary since it's being assigned right back to a nullable variable again.

**Actual Result**

No error.

This happens because _inside_ the scope of the `if`, the type of `y` is narrowed to exclude `undefined`. If there were more code inside that block, it could rely on `y` being non-null without TSC complaining.

However, in cases like this where there isn't any more code inside the block, there's nothing relying on that narrowing -- everything would compile just as successfully with the `!` operator removed.

**Additional Info**

The rule correctly flags simpler examples where there is no temporary type narrowing -- like this one (from #1310):

```ts
let x: string | undefined = Math.random() > 0.5 ? "yes" : undefined;
let y: string | undefined = x!;
```

The more complex example above may be a harder scenario to detect, but it does still fit the bill as "unnecessary," IMHO.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.12.1` |
| `node`                             | `12.18.3` |

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/55111afd7819d29d65da4f41cc6a129f34aaeb3e/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L163

This line asks TS for the contextual type of the non-null expression - which means it is attempting to get the type that TS thinks it's assigning to.
For `let z: string | undefined = x!`, this returns the type `string | undefined`.
However, for `y = x!`, this returns `null` (i.e. ts is telling us it doesn't know).

This is a gap in the handling of our function:
https://github.com/typescript-eslint/typescript-eslint/blob/e7528e686f5fe5cce8504fc15d3cd06b8733712e/packages/eslint-plugin/src/util/types.ts#L477-L512

Should be simple enough to fix up.

**Comment by msquinn** [comment]
@peterflynn  are you working on this? If not I'd love to give it a shot tonight, thanks 😄

**Comment by bradzacher** [comment]
We're going to have to rollback this change.
For simple cases like you presented - sure, it makes sense to validate this the non-null assertion is unnecessary.

However what you may not realise is that this actually changes the control-flow of the rest of the code.

Here is an example:

```ts
let x: number | undefined;
let y: number | undefined;
// typeof y === number | undefined

y = x;
// typeof y === number | undefined

// error: Object is possibly 'undefined'.(2532)
(y + 1);

y = x!;
// typeof y === number     !!!!!!!!

// no error below
(y + 1);
```
